### PR TITLE
NMS-12980: handle all ifOperStatus values in SNMP interface poller

### PR DESCRIPTION
**Pull request by dschlenk**
### All Contributors

* [X] Have you read and followed our [Contribution Guidelines](https://github.com/OpenNMS/opennms/blob/develop/CONTRIBUTING.md)?
* [X] Have you [made an issue in the OpenNMS issue tracker](https://issues.opennms.org)?<br>If so, you should:
  1. update the title of this PR to be of the format: `${JIRA-ISSUE-NUMBER}: subject of pull request`
  2. update the JIRA link at the bottom of this comment to refer to the real issue number
  3. prefix your commit messages with the issue number, if possible
* [X] Have you made a comment in that issue which points back to this PR?
* [X] Have you updated the JIRA link at the bottom of this comment to link to your issue?
* [X] If this is a new or updated feature, is there documentation for the new behavior?
* [X] If this is new code, are there unit and/or integration tests?
* [ ] If this PR targets a `foundation-*` branch, does it avoid changing files in `$OPENNMS_HOME/etc/`?

### Pull Request Process

One or more reviewers should be assigned to each PR.

If you know that a particular person is subject matter expert in the area your PR affects, feel free to assign one or more reviewers when you create this PR, otherwise reviewers will be assigned for you.

If you have made additions or changes to the documentation, or if you _need_ documentation for these code changes, please make sure a technical writer has looked it over.

Once the reviewer(s) accept the PR and the branch passes continuous integration, the PR is eligible for merge.

At that time, if you have commit access (are an OpenNMS Group employee or a member of the Order of the Green Polo) you are welcome to merge the PR.
Otherwise, a reviewer can merge it for you.

Thanks for taking time to contribute!

### External References

* JIRA (Issue Tracker): http://issues.opennms.org/browse/NMS-12980

### CODE REVIEW

**Comment by dschlenk** [comment]
clearly this is a WIP, just hopping in with a PR to get tests to run

**Comment by dschlenk** [comment]
alright I've done some manual testing and found some issues that are now corrected. I also [documented my testing](https://gist.github.com/dschlenk/06f5c29b11bef756ecbb5afa0499cd87).

File: opennms-base-assembly/src/main/filtered/etc/events/opennms.pollerd.events.xml [hunk 1]
```diff
@@ -135,8 +135,91 @@
...
+        <alarm-data reduction-key="uei.opennms.org/nodes/snmp/interfaceOperDown:%dpname%:%nodeid%:%parm[snmpifindex]%" alarm-type="1" auto-clean="false">
+            <update-field field-name="severity" update-on-reduction="true"/>
+        </alarm-data>
+    </event>
+    <event>
+        <uei>uei.opennms.org/nodes/snmp/interfaceOperNotPresent</uei>
+        <event-label>OpenNMS-defined node event: snmp interface Oper Status Not Present</event-label>
+        <descr>&lt;p>The operational status of interface is not present
+            &lt;/p> &lt;p>This event is generated when an snmp poll on interface find the operational status not present.
+            &lt;/p> &lt;p>The notPresent state indicates that the interface is down specifically because
```
Comment (christianpape) [inline opennms-base-assembly/src/main/filtered/etc/events/opennms.pollerd.events.xml:195]: Maybe "The not present state" like in the other definitions.

File: opennms-config-model/src/main/java/org/opennms/netmgt/config/snmpinterfacepoller/Package.java [hunk 2]
```diff
@@ -231,7 +231,7 @@ public boolean equals(final Object obj) {
 
         if (obj instanceof Package) {
             final Package that = (Package)obj;
-            return Objects.equals(this.m_name, this.m_name)
```
Comment (christianpape) [inline opennms-config-model/src/main/java/org/opennms/netmgt/config/snmpinterfacepoller/Package.java:234]: Good catch!

File: opennms-config/src/test/java/org/opennms/netmgt/config/SnmpInterfacePollerConfigFactoryTest.java [hunk 3]
```diff
@@ -0,0 +1,163 @@
+/*******************************************************************************
+ * This file is part of OpenNMS(R).
+ *
+ * Copyright (C) 2006-2014 The OpenNMS Group, Inc.
```
Comment (christianpape) [inline opennms-config/src/test/java/org/opennms/netmgt/config/SnmpInterfacePollerConfigFactoryTest.java:4]: Copyright (C) 2020 The OpenNMS Group, Inc.

File: opennms-doc/guide-admin/src/asciidoc/text/provisioning/examples.adoc [hunk 4]
```diff
@@ -280,6 +280,13 @@ appropriate action. If during the policy evaluation process any policy
 matches for a “DO_NOT_PERSIST” action, no further policy evaluations
 will happen for that particular entity (IP Interface, SNMP Interface).
 
+Another use of this policy is to mark interfaces for polling by the SNMP Interface Poller. The SNMP Interface Poller is a separate daemon that is disabled by default. In order for a this daemon to do any work, some SNMP interfaces need to be selected for polling.
```
Comment (christianpape) [inline opennms-doc/guide-admin/src/asciidoc/text/provisioning/examples.adoc:283]: "In order for a this daemon" -> "In order for this daemon"

File: opennms-doc/guide-admin/src/asciidoc/text/provisioning/examples.adoc [hunk 5]
```diff
@@ -280,6 +280,13 @@ appropriate action. If during the policy evaluation process any policy
...
+Use the "ENABLE_POLLING" and "DISABLE_POLLING" actions available in this policy in order to manage which SNMP interfaces are polled by this daemon.
+Let's create another policy named _pollVoIPDialPeers_ that marks interfaces with _ifType_ _104_. We'll set the _action_ to _ENABLE_POLLING_ and _matchBehavior_ to _ALL_PARAMETERS_.
+Add a parameter for _ifType_ as the key and _104_ as the value.
+
+Decide to move all your meetings to Minecraft and Mumble? Decommission your voice circuits and flip the the action to _DISABLE_POLLING_!
```
Comment (christianpape) [inline opennms-doc/guide-admin/src/asciidoc/text/provisioning/examples.adoc:288]: What?

Comment (christianpape) [reply opennms-doc/guide-admin/src/asciidoc/text/provisioning/examples.adoc:288]: "the the" -> "the"

File: opennms-doc/guide-admin/src/asciidoc/text/snmp-poller/concepts.adoc [hunk 6]
```diff
@@ -0,0 +1,211 @@
+
+// Allow GitHub image rendering
+:imagesdir: ../../images
+
+=== SNMP Interface Poller
+
+_{opennms-product-name}_ can monitor the status of an SNMP interface that is not directly monitored by a <<ga-service-assurance, Service Assurance>> monitor.
+
+The SNMP Interface Poller is a separate service that uses a node interface, typically a management interface, to get administration and operational status for other interfaces present on the device.
+On a configurable time interval an SNMP GET request is made for two specific objects in an `ifEntry` of the `ifTable` defined in the `IF-MIB`:
```
Comment (christianpape) [inline opennms-doc/guide-admin/src/asciidoc/text/snmp-poller/concepts.adoc:10]: a SNMP GET

File: opennms-services/src/main/java/org/opennms/netmgt/snmpinterfacepoller/pollable/PollableSnmpInterface.java [hunk 7]
```diff
@@ -28,23 +28,21 @@
...
-import java.util.Date;
-import java.util.HashMap;
-import java.util.List;
-import java.util.Map;
+import java.util.*;
 
 import org.opennms.netmgt.events.api.EventConstants;
 import org.opennms.netmgt.model.OnmsSnmpInterface;
 import org.opennms.netmgt.poller.PollStatus;
+import org.opennms.netmgt.snmpinterfacepoller.pollable.SnmpInterfaceStatus.*;
```
Comment (christianpape) [inline opennms-services/src/main/java/org/opennms/netmgt/snmpinterfacepoller/pollable/PollableSnmpInterface.java:36]: No .* imports, please.

File: opennms-services/src/main/java/org/opennms/netmgt/snmpinterfacepoller/pollable/PollableSnmpInterface.java [hunk 8]
```diff
@@ -66,75 +64,105 @@
...
-                this.adminstatus = IF_UNKNOWN;
-            else this.adminstatus = adminstatus;
+
+        /**
+         *
+         * @param adminstatus valid values are up(1), down(2), testing(3) according to RFC 2863. Value will be set to
+         *                    invalid(0) in the case where an improper value is attempted to be set.
+         */
+        public void setAdminstatus(SnmpInterfaceStatus adminstatus) {
+            if (adminstatus.getMibValue() > 3 )
```
Comment (christianpape) [inline opennms-services/src/main/java/org/opennms/netmgt/snmpinterfacepoller/pollable/PollableSnmpInterface.java:109]: Please use { ... } for blocks, even for one-line blocks.

**Review by christianpape — CHANGES_REQUESTED** [review]
Looks good. Only some small changes needed.

**Review by christianpape — APPROVED** [review]
Looks good to me.

**Comment by Bonrob2** [comment]
@dschlenk going forward, for documentation changes, please add me to the review. Thanks!

**Comment by dschlenk** [comment]
@Bonrob2 will do! I wrote most of it while stuck in a winnebago in northern MN, was it that obvious?

**Comment by Bonrob2** [comment]
@dschlenk LOL. Not at all, minor tweaks ... it just helps us as we migrate the docs to see what's happening. :-)

**Pull request merged**